File: go/booking-up-for-beauty/booking_up_for_beauty.go
package booking

import (
	"time"
)

// Schedule returns a time.Time from a string containing a date.
func Schedule(date string) time.Time {
	layouts := []string{
		"1/2/2006 15:4:5",
		"January 2, 2006 15:4:5",
		"Monday, January 2, 2006 15:4:5",
		"1/2/2006 15:4:5",
	}
	for _, layout := range layouts {
		t, err := time.Parse(layout, date)
		if err == nil {
			return t
		}
	}
	return time.Time{}
}

// HasPassed returns whether a date has passed.
func HasPassed(date string) bool {
	return Schedule(date).Before(time.Now())
}

// IsAfternoonAppointment returns whether a time is in the afternoon.
func IsAfternoonAppointment(date string) bool {
	return Schedule(date).Hour() >= 12 && Schedule(date).Hour() < 18
}

// Description returns a formatted string of the appointment time.
func Description(date string) string {
	return Schedule(date).Format("You have an appointment on Monday, January 2, 2006, at 15:4.")
}

// AnniversaryDate returns a Time with this year's anniversary.
func AnniversaryDate() time.Time {
	return time.Date(time.Now().Year(), time.September, 15, 0, 0, 0, 0, time.UTC)
}

File: fsharp/booking-up-for-beauty/BookingUpForBeauty.fs
module BookingUpForBeauty

// The following line is needed to use the DateTime type
open System

let schedule (appointmentDateDescription: string): DateTime = 
    DateTime.Parse(appointmentDateDescription)

let hasPassed (appointmentDate: DateTime): bool = 
    DateTime.Compare(appointmentDate, DateTime.Now) < 0

let isAfternoonAppointment (appointmentDate: DateTime): bool =
    appointmentDate.Hour < 18 && 12 <= appointmentDate.Hour

let description (appointmentDate: DateTime): string = 
    appointmentDate.ToString("M/d/yyyy h:mm:ss tt")
    |> sprintf "You have an appointment on %s." 

let anniversaryDate(): DateTime = 
    DateTime()

